engine/consensus/dkg/dkg_controller_test.go

```diff
@@ -0,0 +1,273 @@
+	id       int
+	channels []chan DKGMessage
+}
+
+func (proc *processor) PrivateSend(dest int, data []byte) {
+	proc.channels[dest] <- DKGMessage{Orig: proc.id, Data: data}
+}
+
+// ATTENTION: Normally the processor requires Broadcast to provide guaranteed
+// delivery (either all nodes receive the message or none of them receive it).
+// Here we are just assuming that with a long enough duration for phases 1 and
+// 2, all nodes are guaranteed to see everyone's messages. So it is important
+// to set timeouts carefully in the tests.
+func (proc *processor) Broadcast(data []byte) {
+	for i := 0; i < len(proc.channels); i++ {
+		if i == proc.id {
+			continue
+		}
+		proc.channels[i] <- DKGMessage{Orig: proc.id, Data: data}
+	}
+}
+
+func (proc *processor) Blacklist(node int) {}
+
+func (proc *processor) FlagMisbehavior(node int, logData string) {}
+
+type testCase struct {
+	totalNodes     int
+	phase0Duration time.Duration
+	phase1Duration time.Duration
+	phase2Duration time.Duration
+}
+
+// TestDKGHappyPath tests the controller in optimal conditions, when all nodes are
+// working correctly.
+func TestDKGHappyPath(t *testing.T) {
+	// Define different test cases with varying number of nodes, and phase
+	// durations. Since these are all happy path cases, there are no messages
+	// sent during phase 1 and 2, all messaging is done in phase 0. So we can
+	// can set shorter durations for phase 1 and 2..
+	testCases := []testCase{
+		testCase{totalNodes: 5, phase0Duration: time.Second, phase1Duration: 100 * time.Millisecond, phase2Duration: 100 * time.Millisecond},
+		testCase{totalNodes: 10, phase0Duration: time.Second, phase1Duration: 100 * time.Millisecond, phase2Duration: 100 * time.Millisecond},
+		testCase{totalNodes: 15, phase0Duration: 5 * time.Second, phase1Duration: 2 * time.Second, phase2Duration: 2 * time.Second},
+	}
+
+	// run each test case
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("%d nodes", tc.totalNodes), func(t *testing.T) {
+			testDKG(t, tc.totalNodes, tc.totalNodes, tc.phase0Duration, tc.phase1Duration, tc.phase2Duration)
+		})
+	}
+}
+
+// TestDKGThreshold tests that the controller results in a successful DKG as
+// long as the minimum threshold for non-byzantine nodes is satisfied.
+func TestDKGThreshold(t *testing.T) {
+	// define different test cases with varying number of nodes, and phase
+	// durations
+	testCases := []testCase{
+		testCase{totalNodes: 5, phase0Duration: time.Second, phase1Duration: time.Second, phase2Duration: time.Second},
+		testCase{totalNodes: 10, phase0Duration: time.Second, phase1Duration: time.Second, phase2Duration: time.Second},
+		testCase{totalNodes: 15, phase0Duration: 5 * time.Second, phase1Duration: 2 * time.Second, phase2Duration: 2 * time.Second},
+	}
+
+	// run each test case
+	for _, tc := range testCases {
+		// gn is the minimum number of good nodes required for the DKG protocol
+		// to go well
+		gn := tc.totalNodes - optimalThreshold(tc.totalNodes)
```

Comment:
Not sure I understood why this test case is different than the happy path. Shouldn't we keep this case for when we introduce malicious behaviours?

Reply:
My intention here was to test the **availability** guarantee of the protocol, whereby correct nodes can run the protocol to completion as long as there are less than t offline nodes (where t is the DKG threshold). Being offline is just one example of byzantine behavior. In the logs we can see that offline nodes are being flagged and blacklisted by the running nodes, and that the running nodes still end up with consistent keys.

Reply:
ah right, I missed that only the "good nodes" run the controller. That's an edge scenario indeed, and it actually confirms   the controller is working well during phase 1 (in the happy path, messages only happen in phase 0).